content/6_Advanced/SCC.mdx

```diff
@@ -258,24 +258,231 @@ int main() {
+
+For example, in the formula
+$
+(\lnot x_1 \lor x_2) \land (x_1 \lor \lnot x_2) \land (\lnot x_2 \lor x_3)
+$, the implication graph would have edges $(\lnot x_1 \rightarrow x_2), (\lnot x_2 \rightarrow x_1), (x_1 \rightarrow \lnot x_2), (x_2 \rightarrow \lnot x_1), (\lnot x_2 \rightarrow x_3)$ and ($\lnot x_3 \rightarrow x_2)$. After constructing the graph and finding the SCCs, we check for consistency. Assuming no contradictions are found, we then assign truth values based on the topological order of the SCCs.
+
+This approach ensures that the 2-SAT problem can be solved efficiently, in linear time relative to the number of variables and clauses, leveraging the properties of directed graphs and the power of SCCs to manage logical dependencies and implications systematically. This makes the 2-SAT problem a notable example of a problem that is solvable in polynomial time, despite being a specific case of the generally intractable Boolean satisfiability problem.
+
+## Implementation
+Now, let's implement the entire algorithm for solving the 2-SAT problem using Kosaraju's algorithm. First, we construct the graph of implications $('adj')$ and find all strongly connected components (SCCs). Kosaraju's algorithm efficiently identifies SCCs in $O(n + m)$ time complexity. During the second traversal of the graph, Kosaraju's algorithm visits these SCCs in topological order, allowing us to assign a component number $comp[v]$to each vertex $v$.
+
+Subsequently, to determine the satisfiability of the 2-SAT problem:
+- For each variable $x$, we compare $comp[x]$ and $comp[\lnot x]$. If $comp[x] = comp[\lnot x]$, it 	indicates that both $x$ and $\lnot x$ belong to same SCCs, implying a contradiction.
+- In such cases, we return $false$ to indicate that no valid assignment satisfies the 2-SAT problem.
+
+Below is the implementation of the solution for the 2-SAT problem, utilizing the graph of implications $adj$ and its transpose $adj^T$, where vertices $2k$ and $2k+1$ correspond to variable $k$ and its negation respectively.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+struct TwoSatSolver {
+	int n_vars;
+	int n_vertices;
+	vector<vector<int>> adj, adj_t;
+	vector<bool> used;
+	vector<int> order, comp;
+	vector<bool> assignment;
+
+	TwoSatSolver(int _n_vars)
+	    : n_vars(_n_vars), n_vertices(2 * n_vars), adj(n_vertices),
+	      adj_t(n_vertices), used(n_vertices), order(), comp(n_vertices, -1),
+	      assignment(n_vars) {
+		order.reserve(n_vertices);
+	}
+	void dfs1(int v) {
+		used[v] = true;
+		for (int u : adj[v]) {
+			if (!used[u]) dfs1(u);
+		}
+		order.push_back(v);
+	}
+
+	void dfs2(int v, int cl) {
+		comp[v] = cl;
+		for (int u : adj_t[v]) {
+			if (comp[u] == -1) dfs2(u, cl);
+		}
+	}
+
+	bool solve_2SAT() {
```

Comment:
i think in this case just rename it to `solve()`

Reply:
done